### PR TITLE
planner: refactor a few code of plan cache

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: ref #54057

Problem Summary: planner: refactor a few code of plan cache

### What changed and how does it work?

Add table dirty flags into the plan cache key

No regression

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by AilinKid — APPROVED** [review]

File: pkg/planner/core/plan_cache_utils.go [hunk 1]
```diff
@@ -319,6 +319,49 @@ func NewPlanCacheKey(sctx sessionctx.Context, stmt *PlanCacheStmt) (key, binding
...
+		}
+		if node.Offset != nil {
+			if offset, isParamMarker := node.Offset.(*driver.ParamMarkerExpr); isParamMarker {
+				typeExpected, val := CheckParamTypeInt64orUint64(offset)
+				if !typeExpected {
+					return "", "", false, "unexpected value after LIMIT", err
+				}
+				hash = codec.EncodeUint(hash, val)
+			}
+		}
```
Comment (hawkingrei) [inline pkg/planner/core/plan_cache_utils.go:360]: This code contains some redundant parts and appears to be simplifiable.

File: pkg/planner/core/plan_cache_utils.go [hunk 2]
```diff
@@ -319,6 +319,54 @@ func NewPlanCacheKey(sctx sessionctx.Context, stmt *PlanCacheStmt) (key, binding
...
+			}
+			if node.Offset != nil {
+				if offset, isParamMarker := node.Offset.(*driver.ParamMarkerExpr); isParamMarker {
+					typeExpected, val := CheckParamTypeInt64orUint64(offset)
+					if !typeExpected {
+						return "", "", false, "unexpected value after LIMIT", err
+					}
+					hash = codec.EncodeUint(hash, val)
+				}
+			}
```
Comment (hawkingrei) [inline pkg/planner/core/plan_cache_utils.go:364]: This code contains some redundant parts and appears to be simplifiable.

**Review by hawkingrei — APPROVED** [review]
Other LGTM

File: pkg/planner/core/plan_cache_utils.go (hunk 2 above)
Comment (qw4990) [reply pkg/planner/core/plan_cache_utils.go]: OK, I've simplified it.

**Pull request merged**